README.md
# Notifier

Websocket and HTTP server where authenticated users can subscribe to changes.

# Run quick example

> ./runexample.sh

# API

## Called by app-clients through websocket:

### subscribe_to_events

It opens a websocket connection, accepting a first message in JSON format { "access\_id" : "client1", "timestamp" : 200 }. If succeed, it will keep  open a websocket connection to the client to keep it update on future events (in JSON format) and those that already happened since last_seen. At any moment the notifier can close the connection, in that case it could be becuase it is too busy (too many open connection) or because the user got dropped out. The events themselves are exactly the same json the client would get after the related Get/Post call (including block-querier).

# Example

> wscat -c http://localhost:8080/subscribe_to_events
> connected (press CTRL+C to quit)
> { "access_id" : "client1", "timestamp" : 2000 }
> 

## Called by connectors through HTTP:

### POST push_events

It take a list of events (in JSON format in the request body) to be pushed and it will always return true. Notice that the notifier might actually ignore those events if there's no app-client subscribed interested on those, but this would be transparent to the connector, who will not care whether their notifications are broadcasted or ignored.

# Example

> curl -H "Content-Type: application/json" -X POST -d '[{"accessid":"client1", "created_at": 10, "payload" : "client1 message"},{"accessid":"client2", "created_at": 20, "payload" : "client2 message"}]' localhost:8080/push_events


###  POST drop_user

It force the connection to that user (in JSON format in the request body { "drop\_access\_id" : "client1" }) to be dropped. Notice that even if this call is not made, the notifier could nevertheless drop subscriptions if they get too many or they don't respond to the heartbeat.

# Example

> curl -H "Content-Type: application/json" -X POST -d '{"drop_access_id":"client1"}' localhost:8080/drop_user

# JSON Event formats

{ "accessid" : «access_id», 
  "created\_at" : «timestampt», 
  "payload" : «what the client would receive, that should look the same as it would had get it from an http request»
}



ChangeLog.md
# Changelog for notifier

## Unreleased changes


test/Spec.hs
{-# LANGUAGE TypeFamilies, DataKinds, DeriveGeneric, TypeOperators, OverloadedStrings, ScopedTypeVariables #-}



import           Control.Monad                       (forever, unless)
import           Control.Monad.STM                   (atomically)
import           Control.Concurrent                  (forkIO, killThread, newMVar, threadDelay)
import           Control.Concurrent.Chan
import           Control.Exception                   (bracket, catch, SomeException, handle)
import           Control.Monad.IO.Class              (liftIO)

import           Data.Text                           (Text)
import qualified Data.Text.IO                 as T
import           Data.Aeson                          (decode, encode)
import           GHC.Generics

import           Network.HTTP.Client          hiding (Proxy, responseBody)
import           Network.Socket                      (withSocketsDo)
import           Network.Wai.Handler.Warp            (run)
import qualified Network.WebSockets           as WS

import           Servant
import           Servant.Client
import           Servant.Server

import           Test.Hspec
import           Test.HUnit                          ((@=?))

import           Notifier

testHost :: String
testHost = "localhost"

testPort :: Int
testPort = 8080

testServerUrl :: String
testServerUrl = "http://" ++ testHost ++ ":" ++ show testPort

withNotifierApp :: IO () -> IO ()
withNotifierApp action = do
  bracket
    (forkIO startNotifier)
    killThread
    (const $ do
        waitSome -- wait for server to be up in previous fork
        action)

businessLogicSpec :: Spec
businessLogicSpec = around_ withNotifierApp $ do
  let clients :<|> push :<|> drop = client notifierHttpAPI
  baseUrl <- runIO $ parseBaseUrl $ testServerUrl
  manager <- runIO $ newManager defaultManagerSettings
  let clientEnv = mkClientEnv manager baseUrl
  ch <- runIO $ newChan -- to sync push and clients subscription threads 
  
  describe "get connected clients" $ do
    it "should get an empty list of clients" $ do    
      result <- runClientM clients clientEnv
      result `shouldBe` Right "Connected clients:"

  describe "test websocket subscribe_to_events" $ do
    it "subscribe 2 clients and push 3 events" $ do
      forkIO $ do
        readChan ch -- wait the two websockets clients are subscribed
        readChan ch
        Right "true" <- runClientM (push [event1, event2, event3, event4]) clientEnv
        return ()
      withSocketsDo $ do
        forkIO $ subscribeWebsocket (subscribeApp ch subscriptionInputTest1 [event1]) `shouldReturn` True
        subscribeWebsocket (subscribeApp ch subscriptionInputTest2 [event2]) `shouldReturn` True

-- { "access_id" : "client1", "timestamp" : 2000 }
subscriptionInputTest1 :: SubscriptionInput
subscriptionInputTest1 = SubscriptionInput {
    access_id = "client1" :: Text
  , timestamp = 1000
  }

subscriptionInputTest2 :: SubscriptionInput
subscriptionInputTest2 = subscriptionInputTest1 { access_id = "client2" :: Text }

event1 :: Event
event1 = Event {
    accessid   = "client1" :: Text
  , created_at = 1000
  , payload    = "client1 event" :: Text
  }

event2 :: Event
event2 = Event {
    accessid   = "client2" :: Text
  , created_at = 3000
  , payload    = "client2 event" :: Text
  }

event3 :: Event
event3 = Event {
    accessid   = "client3" :: Text
  , created_at = 4000
  , payload    = "client3 event" :: Text
  }

-- old timestamp
event4 :: Event
event4 = Event {
    accessid   = "client1" :: Text
  , created_at = 500
  , payload    = "old client1 event" :: Text
  }

subscribeWebsocket :: WS.ClientApp Bool -> IO Bool
subscribeWebsocket = WS.runClient "127.0.0.1" testPort "/subscribe_to_events"

subscribeApp :: Chan () -> SubscriptionInput -> [Event] -> WS.ClientApp Bool
subscribeApp ch subs events conn = do
  WS.sendTextData conn (encode subs)
  writeChan ch () -- sync with push event thread
  msg <- WS.receiveData conn
  WS.sendCloseCode conn 1000 ("" :: Text)
  expectCloseException conn "close"
  return $ (decode msg) == Just events

waitSome :: IO ()
waitSome = threadDelay $ 200 * 1000

expectCloseException :: WS.Connection -> Text -> IO ()
expectCloseException conn msg = act `catch` handler
    where
        act = WS.receiveDataMessage conn >>= \ msg -> error $ "Expecting CloseRequest exception, but received:" ++ show msg
        handler (WS.CloseRequest i msg') = do
            i @=? 1000
--            decode utf8 msg' @=? msg
        handler WS.ConnectionClosed     = error "Unexpected connection closed"
        handler (WS.ParseException _)   = error "Unexpected parse exception"
        handler (WS.UnicodeException _) = error "Unexpected unicode exception"        

spec :: Spec
spec = businessLogicSpec

main :: IO ()
main = hspec spec  
    



runexample.sh
#!/bin/sh

stack run &

sleep 2

gnome-terminal -- /bin/sh -c 'wscat -c http://localhost:8080/subscribe_to_events -x "{ \"access_id\" : \"client1\", \"timestamp\" : 10 }" -w 240; sleep 10'

gnome-terminal -- /bin/sh -c 'wscat -c http://localhost:8080/subscribe_to_events -x "{ \"access_id\" : \"client2\", \"timestamp\" : 10 }" -w 240; sleep 10'

gnome-terminal -- /bin/sh -c 'wscat -c http://localhost:8080/subscribe_to_events -x "{ \"access_id\" : \"client3\", \"timestamp\" : 10 }" -w 240; sleep 10'

sleep 2

curl -H "Content-Type: application/json" -X POST -d '[{"accessid":"client1", "created_at": 30, "payload" : "client1 message"},{"accessid":"client2", "created_at": 20, "payload" : "client2 message"}]' localhost:8080/push_events

sleep 2

curl -H "Content-Type: application/json" -X POST -d '[{"accessid":"client3", "created_at": 30, "payload" : "client1 message"}]' localhost:8080/push_events

pkill notifier-exe




src/client.hs
--------------------------------------------------------------------------------
{-# LANGUAGE OverloadedStrings #-}
module Main
    ( main
    ) where


--------------------------------------------------------------------------------
import           Control.Concurrent  (forkIO)
import           Control.Monad       (forever, unless)
import           Control.Monad.Trans (liftIO)
import           Network.Socket      (withSocketsDo)
import           Data.Text           (Text)
import qualified Data.Text           as T
import qualified Data.Text.IO        as T
import qualified Network.WebSockets  as WS


--------------------------------------------------------------------------------
app :: WS.ClientApp ()
app conn = do
    putStrLn "Connected!"

    -- Fork a thread that writes WS data to stdout
    _ <- forkIO $ forever $ do
        msg <- WS.receiveData conn
        liftIO $ T.putStrLn msg

    -- Read from stdin and write to WS
    let loop = do
            line <- T.getLine
            unless (T.null line) $ WS.sendTextData conn line >> loop

    loop
    WS.sendClose conn ("Bye!" :: Text)


--------------------------------------------------------------------------------
main :: IO ()
main = withSocketsDo $ WS.runClient "127.0.0.1" 8080 "/subscribe_to_events" app


html/client.js
function create(n) {
  return () => {
    connections[n-1] = new WebSocket("ws://localhost:8080/subscribe_to_events");

    document.querySelector("#disconnect"+n).addEventListener("click", () => {  connections[n-1].close() });
    document.querySelector("#connect"+n).disabled = true;
    
    connections[n-1].onopen = (event) => {
      console.log("WebSocket client "+n+" is open now.");
      // send connection JSON
      connections[n-1].send(JSON.stringify({ access_id : "client" + n ,timestamp : 2000 }));
      document.querySelector("#disconnect"+n).disabled = false;          
    };

    connections[n-1].onclose = (event) => {
      console.log("WebSocket client "+n+" is closed now.");
      document.querySelector("#disconnect"+n).disabled = true;
      document.querySelector("#connect"+n).disabled = false;      
    };
    
    connections[n-1].onerror = (event) => {
      console.error("WebSocket client "+n+" error observed:", event);
      document.querySelector("#disconnect"+n).disabled = true;
      document.querySelector("#connect"+n).disabled = false;      
    };

    connections[n-1].onmessage = (event) => {
      // append received message from the server to the DOM element 
      const chat = document.querySelector("#events"+n);
      var objs = JSON.parse(event.data);
      objs.forEach(obj => chat.innerHTML += obj.payload + '<br>')
    }
  }
}

var connections = new Array(columns);

var width = 100/columns;

for(i = 1;i<=columns;i++) {
  document.getElementById("parentId").innerHTML += `
		 <div id="clientdiv`+i+`" style="float:left;width:`+width+`%">
		   <h1>Client `+i+`</h1>
		   <button id="connect`+i+`">Connect</button><br>
		   <button id="disconnect`+i+`">Disconnect</button><br>
		   <h2>Events</h2>    
		   <div id="events`+i+`"></div>
		 </div>`;
}

for(i = 1;i<=columns;i++) {		 
  document.querySelector("#connect"+i).addEventListener("click", create(i));		 
  document.querySelector("#disconnect"+i).disabled = true;
}



src/Notifier.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}
{-# LANGUAGE DeriveGeneric     #-}

module Notifier
    ( startNotifier,
      notifierApp,
      newServerState,
      notifierAPI,
      notifierHttpAPI,
      SubscriptionInput(..),
      Event(..)
    ) where

import Prelude              hiding   (concat)

import Control.Monad                 (forever)
import Control.Monad.IO.Class        (MonadIO (..))
import Control.Exception             (finally, catch)
import Control.Concurrent.Async      (mapConcurrently_)
import Control.Concurrent            (MVar, newMVar, modifyMVar_, readMVar)
import Data.Aeson                    (FromJSON, ToJSON, decode, encode)
import Data.Text                     (Text, concat, pack)
import Data.List                     (groupBy)

import GHC.Generics                  (Generic)
import Network.Wai                   (Application)
import Network.Wai.Handler.Warp      (run)
import Servant                       ((:<|>) (..), (:>), Proxy (..), Server, serve, Get, Post, Handler, PlainText, JSON, ReqBody)
import Servant.API.WebSocket         (WebSocket)

import qualified Data.HashMap.Strict        as M
import qualified Data.Text.Lazy             as T (toStrict)
import qualified Data.Text                  as T
import qualified Data.Text.Lazy.Builder     as B
import qualified Data.Text.Lazy.Builder.Int as B
import qualified Data.Text.IO               as T
import qualified Network.WebSockets         as WS

type Timestamp = Int

type Clients = M.HashMap Text (WS.Connection, Timestamp)

data DropUser = DropUser
  {
    drop_access_id :: Text
  } deriving (Eq, Show, Generic)

instance FromJSON DropUser
instance ToJSON   DropUser

data SubscriptionInput = SubscriptionInput
  {
    access_id :: Text
  , timestamp :: Timestamp
  } deriving (Eq, Show, Generic)

instance FromJSON SubscriptionInput
instance ToJSON   SubscriptionInput

data Event = Event 
  {  accessid :: Text
  ,  created_at :: Timestamp
  ,  payload :: Text
  } deriving (Eq, Show, Generic)

instance FromJSON Event
instance ToJSON   Event

-- | Clients handling abstrated =============================

-- | Create a new, initial state:
newServerState :: Clients
newServerState = M.empty

-- | Get client connection
getClient :: Text -> Clients -> Maybe (WS.Connection, Timestamp)
getClient = M.lookup

-- | Add a client (this does not check if the client already exists, you should do
-- | this yourself using `clientExists`):
addClient :: Text -> (WS.Connection, Timestamp)  -> Clients -> Clients
addClient = M.insert

-- | Remove a client:
removeClient :: Text -> Clients -> Clients
removeClient = M.delete

-- | Servant API =======================================================================
type NotifierHttpAPI =
           "clients"     :> Get '[PlainText] Text -- testing endpoint 
      :<|> "push_events" :> ReqBody '[JSON] [Event]  :> Post '[PlainText] Text
      :<|> "drop_user"   :> ReqBody '[JSON] DropUser :> Post '[PlainText] Text

type NotifierWebsocketAPI =      
           "subscribe_to_events" :> WebSocket

type NotifierAPI = NotifierHttpAPI :<|> NotifierWebsocketAPI

server :: MVar Clients -> Server NotifierAPI
server state = (    clientsHttp       state
               :<|> pushEvents        state
               :<|> dropClient        state)
               :<|> subscribeToEvents state

notifierApp :: MVar Clients -> Application
notifierApp = serve notifierAPI . server

notifierHttpAPI :: Proxy NotifierHttpAPI
notifierHttpAPI = Proxy

notifierAPI :: Proxy NotifierAPI
notifierAPI = Proxy

startNotifier :: IO ()
startNotifier = do
  putStrLn "Starting Notifier Server on http://localhost:8080"
  state <- newMVar newServerState
  run 8080 $ notifierApp state

clientsHttp :: MVar Clients -> Handler Text
clientsHttp state = liftIO $ do
  clients <- readMVar state
  return $ "Connected clients:" <> (concat $ map (<> " ") $ M.keys clients)

-- | Drop User
-- | Example: curl -H "Content-Type: application/json" -X POST -d '{"drop_access_id":"client1"}' localhost:8080/drop_user
dropClient :: MVar Clients -> DropUser -> Handler Text
dropClient state (DropUser { drop_access_id = accessid }) = liftIO $ do
  clients <- readMVar state
  case getClient accessid clients of
    Nothing        -> return ()
    Just (conn, _) -> do
      T.putStrLn $ "Droped client: " <> accessid
      WS.sendClose conn ("close" :: Text)
  return "true"

-- | Push Events
-- | Example: curl -H "Content-Type: application/json" -X POST -d '[{"accessid":"ernesto", "created_at": 10, "payload" : "contenido  otroernesto"},{"accessid":"ernesto2", "created_at": 10, "payload" : "contenido otro ernesto2"}]' localhost:8080/push_events
pushEvents :: MVar Clients -> [Event] -> Handler Text
pushEvents state events = liftIO $ do
  T.putStrLn $ "Pushed events: " <> pack (show events)
  clients <- readMVar state
  mapConcurrently_ (\ evs@(e:_) -> case getClient (accessid e) clients of -- TODO: update client timestamp ? With max. timestamp in client's messages 
                                     Nothing            -> return ()
                                     Just (conn, tstmp) -> sendClient conn (filter (\ e -> created_at e >= tstmp) evs))
                   (groupBy (\ e1 e2 -> accessid e1 == accessid e2) events) -- TODO: this only group contiguous events of the same client !!
  return $ "true"

-- | Client subscription
-- | Example: > wscat -c http://localhost:8080/subscribe_to_events
-- |          connected (press CTRL+C to quit)
-- |          > { "access_id" : "client1", "timestamp" : 2000 }
subscribeToEvents :: MonadIO m => MVar Clients -> WS.Connection -> m ()
subscribeToEvents state conn = liftIO . WS.withPingThread conn 30 (return ()) $ do
  msg <- WS.receiveData conn
  case decode msg :: Maybe SubscriptionInput of
    Nothing -> WS.sendTextData conn ("Wrong JSON announcement" :: Text)
    Just (SubscriptionInput{ access_id = user, timestamp = tstmp })  -> do
      -- TODO: missing access_id validation against ON-RAMP server
      T.putStrLn $ "Subscribed client: " <> user 
      flip finally (disconnectClient state user) $ do
        modifyMVar_ state $ \clients -> return $ addClient user (conn, tstmp)  clients
        talkClient conn

talkClient :: WS.Connection -> IO ()
talkClient conn = forever (WS.receiveData conn :: IO Text) `catch` closeConnection -- TODO: messages from clients are meaningless?
  where
    closeConnection (WS.CloseRequest _ _) = T.putStrLn "Connection closed in a good way"
    closeConnection WS.ConnectionClosed   = T.putStrLn "Connection closed in an unpolite way"
    closeConnection e                     = T.putStrLn $ "Something strange happen" <> pack (show e)

sendClient :: WS.Connection -> [Event] -> IO ()
sendClient _    []     = return ()
sendClient conn events = WS.sendTextData conn (encode events)

-- | Remove client and return new state
disconnectClient :: MonadIO m => MVar Clients -> Text -> m ()
disconnectClient state client = 
  liftIO . modifyMVar_ state $ \s -> return $ removeClient client s
  
